Implement Cycle Equivalence analysis

andrealinux1 opened the pull request:
We introduce the `CycleEquivalence` analysis.

This is an analysis which implements the _Cycle Equivalence_ computation algorithm, and provides as result the _cycle equivalence classes_.

Specifically, we introduce:
- The `CycleEquivalenceClass`, which is the unitary element computed by the analysis.
- The `CycleEquivalenceAnalysis`, which contains the algorithm to compute the cycle equivalence.
- The `CycleEquivalencePass`, a `FunctionPass` that can be used to perform the analysis on a `llvm::Function`.

The algorithm is composed by various stages:
- We construct a new `GenericGraph` object, replicating the input CFG, with the addition of the `exit`->`entry` edge.
- Taking advantage of `llvm::GraphTraits<Undirected<>>`, we can now implement the algorithm working on an undirected version of the input.
- We perform the `CycleEquivalence` computation, returning a `llvm::SmallVector` of `CycleEquivalenceClass` objects.

In addition to using the `llvm::GraphTraits<Undirected<>>` traits to walk on the equivalent undirected graph, we also need to:
- Perform a preliminary DFS, in order to:
  - Assign the DFS number to each node in the graph.
  - Compute the spanning tree, and use this information to distinguish tree edges and back edges when running the algorithm.

We also implement the `llvm::DOTGraphTraits` for the `CycleEquivalenceAnalysis<llvm::BasicBlock *>` specialization. In this way, we can have a graphical representation of the undirected graph used internally in the `CycleEquivalenceAnalysis` core implementation.

aleclearmind commented on tests/unit/llvm_lit_tests/CycleEquivalence.ll:
```diff
@@ -0,0 +1,187 @@
+; CHECK: [cycle-equivalence] Bracket Equivalence Class ID: 1
+; CHECK: [cycle-equivalence] while.cond -> while.end
+; CHECK: [cycle-equivalence] entry -> while.cond
+; CHECK: [cycle-equivalence]
```
Make sure this is the minimal thing you need to test so that we don't break tests in case of minor changes.

Tests should not be a copy paste of the current output, but it should be possible to explain each line we're testing .

andrealinux1 replied:
Right, I tried to minimize the test removing, for example, the blank lines and the name of the logger. I also tried to make the checks more robust with the use of  `CHECK-LABEL` and `CHECK-NEXT` in place of the simple `CHECK`, in order to avoid order-related false negatives. This may trigger order-related false positives, but since we are also interested in determinism, I think that we should take a look at possible order-related false positives should they arise

pfez commented on lib/RestructureCFG/CycleEquivalenceClass.cpp:
```diff
@@ -0,0 +1,48 @@
+CycleEquivalenceClass<NodeT>::edge_const_iterator
+CycleEquivalenceClass<NodeT>::end() {
+  return Edges.end();
+}
```
All these should go in the `CycleEquivalenceClass.h` header, otherwise you can't instantiate the template for other `NodeT`s in other `.cpp` files.

andrealinux1 replied:
I now moved all these methods into the header, so we can instantiate another template version from other `.cpp` files. https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/include/revng-c/RestructureCFG/CycleEquivalenceClass.h

pfez commented on lib/RestructureCFG/CycleEquivalenceClass.cpp:
```diff
@@ -0,0 +1,48 @@
+    revng_log(CycleEquivalenceAnalysisLogger,
+              Edge.first->getName()
+                << " -> " << Edge.second->getName() << "\n");
+  }
```
Don't `revng_log` here. If you do it you're constraining all users of the `print` method to print on this logger. Return a `std::string` instead, and leave the decision of where to log to the callers of `print`.

andrealinux1 replied:
I removed all the `revng_log` uses, and noe the `print` method produces a `std::string` here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/include/revng-c/RestructureCFG/CycleEquivalenceClass.h#L36

pfez commented on lib/RestructureCFG/CycleEquivalenceClass.cpp:
```diff
@@ -0,0 +1,48 @@
+
+#include "revng-c/RestructureCFG/CycleEquivalenceClass.h"
+
+extern Logger<> CycleEquivalenceAnalysisLogger;
```
You can drop this, since `print()` shouldn't log.

andrealinux1 replied:
Dropped the logger from the `CycleEquivalenceClass`.

pfez commented on lib/RestructureCFG/CycleEquivalenceClass.cpp:
```diff
@@ -0,0 +1,48 @@
+}
+
+// Explicit template instantiation for the `llvm::BasicBlock *` parameter
+template class CycleEquivalenceClass<llvm::BasicBlock *>;
```
Basically, after the changes I suggested, this compilation unit stays alive only for providing this explicit instantiation. But that isn't necessary anymore since all the methods have been moved to the header. You can just drop this whole compilation unit.

andrealinux1 replied:
This compilation unit has been dropped.

pfez commented on include/revng-c/RestructureCFG/CycleEquivalencePass.h:
```diff
@@ -0,0 +1,32 @@
+
+  void getAnalysisUsage(llvm::AnalysisUsage &AU) const override;
+
+  llvm::SmallVector<CycleEquivalenceClass<llvm::BasicBlock *>> &getResult();
```
Why is this returning a `&` and not a `const &`?

Do we plan to have users of `getResult` modify it?

andrealinux1 replied:
No specific reason, changed it to returning a `const &` here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/include/revng-c/RestructureCFG/CycleEquivalencePass.h#L31.

pfez commented on lib/RestructureCFG/CycleEquivalencePass.cpp:
```diff
@@ -0,0 +1,35 @@
+
+static constexpr const char *Flag = "cycle-equivalence";
+using Reg = llvm::RegisterPass<CycleEquivalencePass>;
+static Reg x(Flag, "Perform the cycle equivalence analysis");
```
Capital `X` please.

Also the description string will show up in the flamegraph of our tracing infrastructure. Please be more concise and just put the pass name: `"Cycle Equivalence Pass"` is ok.

andrealinux1 replied:
I applied these changes here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalencePass.cpp#L12.

pfez commented on lib/RestructureCFG/CycleEquivalencePass.cpp:
```diff
@@ -0,0 +1,35 @@
+  CycleEquivalenceAnalysis<llvm::Function *> CEA;
+  CEA.run(&F);
+
+  EquivalenceClasses = CEA.getCycleEquivalenceClasses();
```
The current design of `getCycleEquivalenceClass` as a method of `CycleEquivalenceAnalysis` makes it a little bit annoyting to set up.

I'd just make `getCycleEquivalenceClasses` a standalone function that takes a `llvm::Function` as argument.
Then internally `getCycleEquivalenceClasses` sets up `CycleEquivalenceAnalysis` and calls `run.

This way, the top-level API for getting the cycle equivalence classes is stateless from the standpoint of callers, that don't need to manage `CycleEquivalenceAnalysis` at all.

andrealinux1 replied:
I rearranged this in a standalone function which instantiates the `CycleEquivalenceAnalysis` and returns directly the results here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalencePass.cpp#L16.

The standalone function is a template function defined here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L498 for which I instantiate a version taking a `llvm::Function *` as a parameter.

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,113 @@
+  void print();
+
+  llvm::SmallVector<CycleEquivalenceClass<NodeT>, 1>
+  getCycleEquivalenceClasses();
```
See my other comment, make this a standalone function that accepts a `llvm::Function *` and sets up `CycleEquivalenceAnalysis` internally.

Also, with recent template magic `SmallVector` can infer the best small size (unfortunately `SmallSet` can't), so just drop the second template argument in the return type.

andrealinux1 replied:
This is now a standalone function defined here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L192.

Also the `SmallVector` second template argument is now omitted.

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,113 @@
+  using BracketDescriptor = std::pair<EdgeDescriptor, size_t>;
+
+private:
+  llvm::SmallVector<CycleEquivalenceClass<NodeT>, 1> CycleEquivalenceClasses;
```
Drop `, 1`. Now `SmallVector` can infer the size.

andrealinux1 replied:
Dropped the useless template parameter https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L131.

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,113 @@
+public:
+  CycleEquivalenceAnalysis() {}
+
+  void insertEdge(EdgeDescriptor E, BracketDescriptor BD);
```
This does not need to be `public` AFAIU.

andrealinux1 replied:
Right, I changed some of these methods as `private`s https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L140.

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,113 @@
+  }
+};
+
+template<class GraphT, class GT = llvm::GraphTraits<GraphT>>
+class CycleEquivalenceAnalysis {
```
Please, add some doxygen to this template class, explaining what it does, and linking the paper.

andrealinux1 replied:
I added some doxygen here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L87

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,468 @@
+  // For every BasicBlock in F, we create a corresponding node in the
+  // `GenericGraph`, and we keep a correspondence between the `BasicBlock` and
+  // the node in the `GenericGraph`
+  auto NodesRange = llvm::make_range(GT::nodes_begin(F), GT::nodes_end(F));
+  for (NodeT BB : NodesRange) {
```
Doesn't `llvm::nodes(F)` work?

andrealinux1 replied:
Yeah, I changed it to use `llvm::nodes` here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L319, but I am a bit conflicted about it. This method (and the class indeed), takes a second template parameter, being the `GraphTraits` telling how to iterate over the original graph. Using `llvm::nodes` prevents this opportunity, while the previous one explicitly used them (as it is done when iterating over the `children`, cfr. here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L338.

pfez replied:
Graph traits, even filtered graphs, or with `Inverse` or `Undirected` never change the view of nodes. They only change how edges are seen.
So you only ever iterate on all nodes, which I think covers your concerns.

Does this make sense?

andrealinux1 replied:
Oh right, this makes all the sense, unless with have some filtered stuff `nodes` should never be impacted by the absence of use of the `GT` parameter.

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,468 @@
+      LabelEdgePool++;
+    }
+
+    // TODO: the `exit` -> `entry` backedge insertion should be performed in the
+    //       agnostic part
+    // If the BasicBlock has no successors, we insert an additional and
+    // artificial backedge to the entry node
+    if (SuccessorsRange.empty()) {
+
+      // We should check that a single exit node has been identified at this
+      // point
+      revng_assert(not ExitConnected);
+
+      From->addSuccessor(EntryNode, 0);
+      ExitConnected = true;
+    }
```
Leave the the handling of exit -> entry edge at the very end, and out of this loop.

Drop `ExitConnected` and adopt a `SmallSet<NodeT>`, where you put all the nodes without successors.

At the very end, after the loop, you just check that the size of the `SmallSet`is exactly 1. If it's more than 1 the graph has many exits, if it's 0 (that was currently unchecked) the graph doesn't have an exit, which is also wrong. In both cases you can `revng_abort` with a meaningful message, so that whoever is using the API can adjust their graph before computing the cycle equivalence classes.

andrealinux1 replied:
I dropped the `ExitConnected` flag, and went with the solution with the `SmallSet` as you suggested here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L358.

andrealinux1 replied:
With the latest introduced changes, and the possibility of sinking graphs with multiple exit, an explicit handling depending on the number of identified exit nodes has been here https://github.com/revng/revng-c/blob/32e8b34c86eae349bd0b23530e69160e8d8b8b43/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L430

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,468 @@
+  using udf_iterator = llvm::df_iterator<BlockNode *, ExtType, false, UGT>;
+
+  BlockNode *EntryNode = Graph.getEntryNode();
+  auto Begin = udf_iterator::begin(EntryNode);
```
`s/Begin/It`

andrealinux1 replied:
Changed this here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L106

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,113 @@
+  std::map<size_t, BracketDescriptor> ClassToBracketDescriptor;
+
+  std::map<BlockNode *, size_t> NodeDFSNumMap;
+  std::map<size_t, BlockNode *> DFSNumNodeMap;
```
This can be a `std::vector<BlockNode *>` or a `SmallVector<BlockNode *>`.

andrealinux1 replied:
This field is now not a part of the `CycleEquivalenceAnalysis` class anymore, but is still needed here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L169 in the `insertCappingBackedges` method.

This is due to the fact that the logic which computes the target of the capping backedge (when we need to insert one), uses merely the DFS index of each node in the computations. Therefore, at the end of the _routine_, we need to go back from the index to the node itself. I now populate this map (which is local to the method) opportunistically each time I have a node available (and can therefore access its DFSNum). The alternative would be to keep a pair of DFSNum and the corresponding node during the algorithm, but that solution would be strictly more memory consuming than the current approach.

pfez replied:
If you keep it as a local variable like you do now it's fine

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,468 @@
+                                                                size_t &HI0,
+                                                                size_t &HI2) {
+
+  // 1) Compute the highest reached node from a backedge starting from the
+  //    current `Block`
```
I believe "highest" here "closer to the spanning tree root". Please clarify.

andrealinux1 replied:
Elaborated more this here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L164

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,468 @@
+}
+
+/// Helper function to compute the capping backedges needed values
+template<class GraphT, class GT>
+void CycleEquivalenceAnalysis<GraphT, GT>::computeCappingValues(BlockNode *BB,
+                                                                size_t &HI0,
+                                                                size_t &HI2) {
```
Make this return a `std::pair<size_t, size_t>` or even better, a plain `struct` with a meaningful name, and 2 `size_t` fields.

andrealinux1 replied:
This method has been dropped entirely, to I think this is updated, the two parameters are now simply local variables in this https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L162C44-L162C66 method

pfez replied:
:+1:

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,113 @@
+public:
+  EdgeIndex(size_t Index) { this->Index = Index; }
+  size_t getValue() const { return Index; }
+  bool operator<(const EdgeIndex &Rhs) const { return Index < Rhs.getValue(); }
+  bool operator==(const EdgeIndex &Rhs) const {
+    return Index == Rhs.getValue();
+  }
```
```c++
bool operator==(const EdgeIndex &) = default;
std::strong_ordering operator<=>(const EdgeIndext &) = default;
```

andrealinux1 replied:
Changed this here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L51

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,113 @@
+
+public:
+  EdgeIndex(size_t Index) { this->Index = Index; }
+  size_t getValue() const { return Index; }
```
Just `value()`.

andrealinux1 replied:
With the new refactoring, storing the needed information directly on the edge, we don't need anymore this field in the associated `Label`, so `value()` has been dropped entirely.

andrealinux1 replied:
After the latest changes, an unique incremental identifier is part of the edge. Here https://github.com/revng/revng-c/blob/32e8b34c86eae349bd0b23530e69160e8d8b8b43/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L44 is the field

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,468 @@
+    BlockNode *SourceNode = Begin.getPath(VisitStackSize - 2);
+
+    // Search for the edge which connects `Source` and `CurrentNode`
+    for (auto &[Neighbor, Index] :
+         llvm::children_edges<llvm::Undirected<BlockNode *>>(SourceNode)) {
+
+      // Check if the successor is equal to `CurrentNode`, and in case add it to
+      // the spanning tree
+      if (Neighbor == CurrentNode) {
+        SpanningTreeEdges.insert(Index);
+        break;
+      }
+    }
```
```c++
auto ChildrenEdgesRange = llvm::children_edges<llvm::Undirected<BlockNode *>>(SourceNode);
auto EdgeIt = llvm::find(ChildrenEdgesRange, {Neighbor, Index});
revng_assert(EdgeIt != ChildrenEdgesRange.end());
SpanningTreeEdges.insert(EdgeIt->second);
```

andrealinux1 replied:
Reimplemented this using your suggestion here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L136.

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,113 @@
+  using BlockGraph = GenericGraph<BlockNode>;
+  using EdgeDescriptor = revng::detail::EdgeDescriptor<BlockNode *>;
+  using EdgesVectorType = llvm::SmallVector<EdgeDescriptor>;
+  using BracketDescriptor = std::pair<EdgeDescriptor, size_t>;
```
What does the `size_t` represent in the pair?

andrealinux1 replied:
It represents the size of the stack of open brackets in the moment the `BracketDescriptor` is created. The pair together represents the equivalence class. I added here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L124 a comment explaining better what this field representes.

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,468 @@
+  for (EdgeDescriptor ForwardEdge : getForwardEdges(BB)) {
+    BlockNode *Target = ForwardEdge.second;
+    HI1 = std::min(HI1, HighestReachedDFSNum.at(Target));
+  }
```
If `Target` is the same as `BB` I think `HighestReachedDFSNum.at(Target)` crashes. Isn't it?

andrealinux1 replied:
I think the reasoning behind your comment is misguided. Indeed, `Target` are not the nodes reached by the backedges starting from `BB`, but are nodes reached by forward edges from BB. Those nodes are guaranteed to have been already processed (since we are working in postorder), and therefore the corresponding entry in `HighestReachedDFSNum` for `Target` is populated (`HighestReachedDFSNum` is a "global" object shared between all the iterations).

Your comment was anyway spot-on, because I think that you were trying to say something like "what happens with a self-loop?". As elaborated above, this is not a bug here, but thinking about it I noticed that self-loops edge did not end up in any equivalence class for other reasons, and I therefore fixed such behavior.

pfez replied:
Yeah you rephrased it better.

In any case this wasn't a bug. Closing this conversation.

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceClass.h:
```diff
@@ -0,0 +1,38 @@
+
+  size_t getID();
+
+  void insertEdge(EdgeDescriptor &E);
```
Just `insert`?

andrealinux1 replied:
Changed this here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/include/revng-c/RestructureCFG/CycleEquivalenceClass.h#L31

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,113 @@
+  std::map<size_t, BracketDescriptor> ClassToBracketDescriptor;
+
+  std::map<BlockNode *, size_t> NodeDFSNumMap;
+  std::map<size_t, BlockNode *> DFSNumNodeMap;
```
These should become per-node members, that are populated by the DFS, as per our in-person discussion

andrealinux1 replied:
These fields have been dropped, due to the bigger refactoring we discussed.

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,113 @@
+
+  // This set is used to store the edges belonging to the spanning tree that we
+  // compute
+  llvm::SmallSet<EdgeLabel, 4> SpanningTreeEdges;
```
These should become per-edge members, that are populated by the DFS, as per our in-person discussion

andrealinux1 replied:
Right, also this field disappeared.

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,113 @@
+private:
+  llvm::SmallVector<CycleEquivalenceClass<NodeT>, 1> CycleEquivalenceClasses;
+
+  std::map<BracketDescriptor, size_t> BracketToClass;
+  std::map<size_t, BracketDescriptor> ClassToBracketDescriptor;
```
The brackets should be added as actual edges on the graph, as per our in-person discussion

andrealinux1 replied:
Mmm wait, I think you misunderstood some detail here.

This map is just used to keep a map between the n-th equivalence class, and the `BracketDescriptor` that gave origin to it. E.g., say we at some point create a new equivalence class, which is the first one, with some edges in it, described by the bracket `(a,1)`, we note down that the 0-th equivalence class is originated by such `BracketDescriptor`.

This information is used only when printing debug information, but I would be very sad removing it, since this information is very useful to debug the internal functioning of  `CycleEquivalenceAnalysis` in case of bugs.

pfez replied:
Mmh ok.

But AFAIU `ClassToBracketDescriptor` is used just for debug prints, which is fine for me but it's worth a comment to say it.

andrealinux1 replied:
Yeah, you understood it correctly, at the moment this map is used just to keep track of the original extensive name of the `bracket` which gave origin to a specific equivalence class. The algorithm per se, doesn't need this mapping, but it is useful for debug prints in case of bug hunting. Added a comment explaining this here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L201

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,113 @@
+  llvm::SmallSet<EdgeLabel, 4> SpanningTreeEdges;
+
+  // Highest reached DFSNum for each `BB`
+  std::map<BlockNode *, size_t> HighestReachedDFSNum;
```
This is only used for computing the capping backedges, so it should be local to the function that in the future will materialize the capping backedges on the graph.

andrealinux1 replied:
This is now the case here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L166

Commit:
CycleEquivalence: implement analysis

We introduce the `CycleEquivalence` analysis.

This is an analysis which implements the _Cycle Equivalence_ computation
algorithm, and provides as result the _cycle equivalence classes_
Specifically, we introduce:
- The `CycleEquivalenceClass`, which is the unitary element computed by
  the analysis.
- The `CycleEquivalenceAnalysis`, which contains the algorithm to
  compute the cycle equivalence.
- The `CycleEquivalencePass`, a `FunctionPass` that can be used to
  perform the analysis on a `llvm::Function`.

The algorithm is composed by various stages:
- We construct a new `GenericGraph` object, replicating the input CFG,
  with the addition of the `exit`->`entry` edge.
- Taking advantage of `llvm::GraphTraits<Undirected<>>`, we can now
  implement the algorithm working on an undirected version of the input.
- We perform the `CycleEquivalence` computation, returning a
  `llvm::SmallVector` of `CycleEquivalenceClass` objects.

In addition to using the `llvm::GraphTraits<Undirected<>>` traits to
walk on the equivalent undirected graph, we also need to:
- Perform a preliminary DFS, in order to:
  - Assign the DFS number to each node in the graph.
  - Compute the spanning tree, and use this information to distinguish
    tree edges and back edges when running the algorithm.

The internal graph used by the analysis also normalizes the graph in
order to have a sinle exit node (called sink), which is a requirement
for the `CycleEquivalence` algorithm.

We also implement the `llvm::DOTGraphTraits` for the
`CycleEquivalenceAnalysis<llvm::Function *>` specialization. In this
way, we can have a graphical representation of the undirected graph used
internally in the `CycleEquivalenceAnalysis` core implementation.

We add some `FileCheck` tests on some well-known graph topologies.

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,554 @@
+    NakedEdge = NakedEdgeDescriptor(E.first->getBlock(), E.second->getBlock());
+  } else {
+    NakedEdge = NakedEdgeDescriptor(E.second->getBlock(), E.first->getBlock());
+  }
```
I'd go for a method on `EdgeDescriptor` so that you can do: `E.getNaked()` and get the `NakedEdgeDescriptor` out.

andrealinux1 replied:
In reading this comment, take into consideration also the snippet highlighted by the next comment here https://github.com/revng/revng-c/pull/18#discussion_r1634367342.

I think there might be a misunderstanding. `EdgeDescriptor` is not a class per se, but it is just an "elegant" way to refer to a `std::pair` of nodes. In this context, when we want to pass from a `EdgeDescriptor` to a `NakedEdgeDescriptor`, the focus is on the fact that we want to unwrap the internal node, in this case in order to get the original `NodeT` which referred to the nodes on the original graph, and not the node that we use on the `GenericGraph` which wraps it with other additional information. Therefore, I don't get how implementing a method on the `EdgeDescriptor` can help us unwrap from the element of the pair (and therefore dropping the `getBlock` method), and how it can help us get the `NakedEdgeDescriptor` describer by the inverted edge or not.

pfez replied:
Ok you can't do a method, but it's disturbing to have this `NakedEdgeDescriptor` because it makes it seem like it's somehow different from a regular `EdgeDescriptor`.
Drop the `Naked` prefix, just do:
```c++
using EdgeDescriptor = CycleEquivalenceClass<NodeT>::EdgeDescriptor;
```

andrealinux1 replied:
Ok, I applied your proposed change [here](https://github.com/revng/revng-c/blob/f23593da9f15288aca2fee41975e6e55dbbbcb3b/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L46)

There is [another place](https://github.com/revng/revng-c/blob/f23593da9f15288aca2fee41975e6e55dbbbcb3b/include/revng-c/RestructureCFG/CycleEquivalenceResult.h#L32) where we use the `NakedEdgeDescriptor`. However, there we need both the `EdgeDescriptor` and `NakedEdgeDescriptor` in the same namespace, so we can't just remove the `Naked` prefix. I think it is legit such use of it there

pfez replied:
Ok, but like I said `Naked` is weird. Basically it means that the edge refers to the original nodes. What about `OriginalNodesEdgeDescriptor`?

andrealinux1 replied:
Sure, changed the name [here](https://github.com/revng/revng-c/blob/b69cf3baad6602473085be7a84750763da0e2b02/include/revng-c/RestructureCFG/CycleEquivalenceResult.h#L34)

andrealinux1 replied:
Also changed the name variable from `NakedEdge` to `OriginalEdge` https://github.com/revng/revng-c/blob/b69cf3baad6602473085be7a84750763da0e2b02/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L46

pfez replied:
:+1:

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,184 @@
+    size_t DFSNum = SizeTMaxValue;
+
+  public:
+    NodeT getBlock() const { return Block; }
```
This is used only for initializing a `NakedEdgeDescriptor`. If you do like I suggested in its only use, this is not necessary anymore.

andrealinux1 replied:
I've a comment on this, which I'll merge together with the comment above. cfr. https://github.com/revng/revng-c/pull/18#discussion_r1634366830.

pfez replied:
Ok forget about this.

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,249 @@
+};
+
+// `struct` used to represent the unique incremental `Index` that is
+// associated with each undirected edge in the `GenericGraph`
```
This is really hard to understand for someone who doesn't already know the design.

I think it's better to expand this comment in a full explanation of why the `EdgeLabel` is designed like it is.

andrealinux1 replied:
I improved the description of `EdgeLabel` in order to explain all the fields of the `struct` and what it is used for here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L30

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,249 @@
+
+public:
+  EdgeLabel(EdgeKind Kind, EdgeType Type, size_t ID, size_t SuccNum) :
+    Kind(Kind), Type(Type), ID(ID), SuccNum(SuccNum) {}
```
Missing empty line after constructor.

andrealinux1 replied:
Added an empty line here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L64

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,249 @@
+  struct BlockBidirectionalNode {
+    BlockBidirectionalNode(NodeT Block) : Block(Block) {}
+
+    // This is used to store a pointer to the wrapped `llvm::BasicBlock`, except
```
Wait, we don't necessarily have `llvm::BasicBlock`s here. I guess you mean `NodeT`?

andrealinux1 replied:
Right, generalized this in the comment here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L149

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,659 @@
+    ClassToBracketDescriptor.emplace(EquivalenceClassIndex, BD);
+    auto &EClass = CycleEquivalenceClasses.emplace_back(EquivalenceClassIndex);
+    EClass.insert(NakedEdge);
+  }
```
The last 2 lines of the `then` and `else` can be factored out after the `if`.

andrealinux1 replied:
I factored these two lines in here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L69

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,249 @@
+public:
+  CycleEquivalenceAnalysis() {}
+
+  CycleEquivalenceClassVector getCycleEquivalenceClasses();
```
Are we even using this now?
Or at least we plan to use it somehow?
Or are we just going to use `getEdgeToCycleEquivalentClassIDMap()`?

pfez replied:
There are various occurrences in the rest but they seem not to be called from anywhere. I'll tag them with other comments.

andrealinux1 replied:
I removed all the implementations and declarations for `getCycleEquivalenceClasses` in ce352f2912e639e1ef307bb1d5b131db48444878

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,249 @@
+
+template<class GraphT>
+CycleEquivalenceAnalysis<GraphT>::CycleEquivalenceClassVector
+getCycleEquivalenceClasses(GraphT F);
```
Other instance of `getCycleEquivalenceClasses`. Seems unused.

andrealinux1 replied:
Now absent in ce352f2912e639e1ef307bb1d5b131db48444878

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,659 @@
+CycleEquivalenceAnalysis<GraphT, GT>::CycleEquivalenceClassVector
+CycleEquivalenceAnalysis<GraphT, GT>::getCycleEquivalenceClasses() {
+  return CycleEquivalenceClasses;
+}
```
Other instance of `getCycleEquivalenceClasses`. Seems unused.

andrealinux1 replied:
Now absent in ce352f2912e639e1ef307bb1d5b131db48444878

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,659 @@
+  CEA.run(F);
+
+  return CEA.getCycleEquivalenceClasses();
+}
```
Other instance of `getCycleEquivalenceClasses`. Seems unused.

andrealinux1 replied:
Now absent in ce352f2912e639e1ef307bb1d5b131db48444878

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,659 @@
+}
+
+template CycleEquivalenceAnalysis<Function *>::CycleEquivalenceClassVector
+getCycleEquivalenceClasses(Function *F);
```
Other instance of `getCycleEquivalenceClasses`. Seems unused.

andrealinux1 replied:
Now absent in ce352f2912e639e1ef307bb1d5b131db48444878

pfez commented on lib/RestructureCFG/CycleEquivalencePass.cpp:
```diff
@@ -0,0 +1,34 @@
+// Copyright rev.ng Labs Srl. See LICENSE.md for details.
+//
+
+#include "revng-c/RestructureCFG/CycleEquivalenceAnalysis.h"
```
Isn't it enough to include `CycleEquivalenceResult.h`?

andrealinux1 replied:
Yes, replaced this here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/lib/RestructureCFG/CycleEquivalencePass.cpp#L6

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,249 @@
+  TreeEdge,
+  BackEdge,
+  Invalid,
+};
```
I'd call the first `OriginalEdgeKind` and the second `SpanningTreeEdgeKind`. The current names are too similar, too easy to confuse.

andrealinux1 replied:
Renamed the fields here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L18 and here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L24

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,659 @@
+    };
+    auto EdgeIt = llvm::find_if(ChildrenEdgesRange, FindCurrentNode);
+    revng_assert(EdgeIt != ChildrenEdgesRange.end());
+    EdgeIt->Label->setType(EdgeType::TreeEdge);
```
I'm dubious about this part. It's not that it doesn't do what we want, but it does it by accident, or at least it heavily relies on implementation details of `Undefined`, which I'm fine with, but we should put it in a comment, otherwise is confusing and subject to breakage in future.

Let me explain.

Like you say in another comment, we should be able to detect precisely the (undirected) edge that brought us here during the DFS. However, in `FindCurrentNode`, you're effectively ignoring the edge and only considering the neighbor, which isn't obviously correct. In fact it seems wrong, but it works for two subtle reasons:
1. because in the `Undirected` the edges that were originally "forward" are always traversed first even in the undirected DFS, so if you do `find_if` on the `ChildrenEdgesRange` (which is a `vector` under the hood) you always stop on the first, which is obviously what actually took you there the first time in the DFS.
2. Even in the case where there in the original graph there are many "forward" edges from A to B, the DFS only ever explores the first, because for all the following cases B has already been visited and the DFS doesn't continue on B. So it never happens that the search of edges we're doing really wants to see any other edge beyond the first.

So yes, this part does what's expected, but from the body of `FindCurrentNode` seems broken, and the reason it's not broken is implementation detail.

Like I said, it's fine, but please put a big comment around it with the information I provided here.

andrealinux1 replied:
Oh yeah sure, I didn't think this through, but I double checked your reasoning and I'm convinced too of your conclusion stating that "by accident" the implementation detail is doing the right thing.

I added a big comment explaining all the above here https://github.com/revng/revng-c/blob/cb8faa8b47f4e6f1d55c07bcf6562f668884c042/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L153

pfez replied:
Great!!!

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,659 @@
+  size_t SourceDFSNum = std::get<0>(ED)->getDFSNum();
+  size_t TargetDFSNum = std::get<1>(ED)->getDFSNum();
+
+  // The equal is very important, or we exclude the self loop backedges
+  return TargetDFSNum <= SourceDFSNum;
```
Why do we exclude self-loop backedges?

pfez replied:
I trust this is the right decision since there's a unit test on self-loops. But it's not clear why. Please improve the comment you say it's `very important` but you don't explain why.

andrealinux1 replied:
I added a more verbose explanation here https://github.com/revng/revng-c/blob/b5fe805f75dde9c9ada264ace247d4b989c4f2d7/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L108.

Basically, the point is that a selfloop source and target nodes DFSNum is equal by defintion. If the equal is left out from the comparison, therefore a selfloop will not satisfy the requirement, and will not be detected as a backedge (while, indeed, it is always a backedge).

pfez replied:
:+1:

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,659 @@
+
+  for (auto &[Neighbor, Label] :
+       llvm::children_edges<llvm::Undirected<BlockNode *>>(Source)) {
+    BlockEdgeDescriptor Edge = BlockEdgeDescriptor({ Source,
```
I read the typename twice on this line. Use `auto`.

andrealinux1 replied:
Changed it to `auto` here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L320C17-L320C36

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,659 @@
+
+  for (auto &[Neighbor, Label] :
+       llvm::children_edges<llvm::Undirected<BlockNode *>>(Source)) {
+    BlockEdgeDescriptor Edge = BlockEdgeDescriptor({ Neighbor,
```
I read the typename twice on this line. Use auto.

andrealinux1 replied:
Changed it to `auto` here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L363

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,659 @@
+
+  for (auto &[Neighbor, Label] :
+       llvm::children_edges<llvm::Undirected<BlockNode *>>(Source)) {
+    BlockEdgeDescriptor Edge = BlockEdgeDescriptor({ Source,
```
I read the typename twice on this line. Use auto.

andrealinux1 replied:
Changed it to `auto` here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L285

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,659 @@
+    BlockEdgeDescriptor Edge = BlockEdgeDescriptor({ Source,
+                                                     Neighbor,
+                                                     Label->id() });
+    if (Label->type() == EdgeType::BackEdge and isBackedge(Edge)) {
```
Can it ever happen that the LHS of `and` is `true` and the RHS `false`??

Or should `isBackedge(Edge)` be an assertion inside the `if`?

pfez replied:
Actually this happens for all the other uses of `isBackedge`. It seems to me that the use of `isBackedge` (or `not isBackedge`) in other occurrences is redundant and should be turned into an assertion.

If not, then there's something I'm missing.

pfez replied:
Perhaps the things I'm missing have to do with self-loops?

andrealinux1 replied:
The terminology is not clear and may lead to confusion, but the LHS and the RHS refers to two very different things. The LHS refers to the fact that the edge on the undirected graph has been marked as a backedge on the spanning tree. The RHS (the invocation to the `isBackedge` helper function) instead, just checks the `DFSNum`s, and tells that a specific edge is a backedge if the `DFSNum` of the source node is greater than the `DFSNum` of the target.

Both the criteria are needed, even if they may seem redundant, and this is due to the fact that on directed graph we can have tree edges, back edges, forward edges and cross edges, while on undirected graphs we can simply have tree edges and back edges. Due to this, the detection of forward and backedges needs both these criteria.

However, with the latest changes here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L321
the check highlights more clearly the distinction between the criteria

pfez replied:
Oh ok it's much clearer now.

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,659 @@
+      }
+    }
+
+    // 5) This is the highest reached node for any child of the current `Block`,
+    //    excluding the one which reaches the highest target
+    size_t HI2 = SizeTMaxValue;
+    for (const auto &[ForwardEdge, _, __] : getForwardEdges(BB)) {
+      BlockNode *Target = std::get<1>(ForwardEdge);
+      if (Target != HI1Child) {
+        HI2 = std::min(HI2, HighestReachedDFSNum.at(Target));
+      }
+    }
+
+    // Insert the capping backedge, if the second highest reached node for any
+    // child of the current `BB`, is higher than any backedges which starts from
+    // `BB`. The capping backedge goes from `BB` to such second highest reached
+    // node.
+    if (HI2 < HI0) {
+      BlockNode *HI2Node = DFSNumNodeMap.at(HI2);
+
```
The only reason why you need `DFSNumNodeMap` is to know the node with `DFSNum == HI2`. But you've just computed `HI2`, and you knew exactly what that node was.

So, drop `DFSNumNodeMap` an rewrite the loop at point `5)` as:
```c++
    size_t HI2 = SizeTMaxValue;
    BlockNode *HI2Node = nullptr;
    for (const auto &[ForwardEdge, _, __] : getForwardEdges(BB)) {
      BlockNode *Target = std::get<1>(ForwardEdge);
      if (Target != HI1Child) {
        size_t NewMin = std::min(HI2, HighestReachedDFSNum.at(Target));
        if (NewMin < HI2)
          HI2Node = Target;
        HI2 = NewMin;
      }
    }
```

andrealinux1 replied:
Mmmmm, I think you got the functioning of this wrong. When I compute the HI2 `DFSNum` (the number of the second highest reached node for each child of the current `BB` node), HI2 does not refer to the `Target` nodes on which I iterate on.

The procedure works like this:
- I iterate on the forward edges from `BB`. The other "end" of such edges will be the `Target` node. Then, using the `HighestReachedDFSNum` mapping, I retrieve the "highiest" `DFSNum` that is reached from each subtree starting from `Target` which has been previously computed. Such highest `DFSNum` does not correspond to `Target`, but can be a node which is arbitrarily "high" in the tree (in other words, there is a "preliminary" phase which for each node, computes the highest reached node by a backedge, which populates the `HighestReachedDFSNum` mapping. Then, in this phase, for each child `Target` of the current `BB`, I select the highest reached `DFSNum` using the mapping).
- The `DFSNum` is then passed in the `DFSNumNodeMap` in order to retrieve the node to which we need to add the capping backedge to.

We can get rid of  `DFSNumNodeMap`, at the cost of saving directly the pointer to `HI2Node` together with the `HI2` `DFSNum` in the `HighestReachedDFSNum`. Let me know if you would prefer things to be done in such way.

pfez replied:
Yes please

andrealinux1 replied:
I implemented the proposed changes from [here](https://github.com/revng/revng-c/blob/f23593da9f15288aca2fee41975e6e55dbbbcb3b/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L211) (the main change is that now the `HighestReachedDFSNum` contains both the `DFSNum` (which is needed to apply the `<` comparisons), and the pointer to the original node.

I think the code is now even less understandable (since it deviates from the paper description), so I also added some comments before each of the 4 steps of the capping backedges introduction, to better explain what the variables represents, hopefully making it easier to reconstruct the functioning of this procedure.

pfez replied:
I checked it out. I kinda like it, but I wouldn't use a `std::pair` in the map, I'd use a custom `struct` locally defined in the function with fields with explanatory naming.

Besides that I'd say the current implementation is fine even if it deviates from the paper

andrealinux1 replied:
Oh ok, I'll use a custom `struct` in place of the `pair`

andrealinux1 replied:
Ok, [here](https://github.com/revng/revng-c/blob/09b768691c58208a87bb2184f9741d98110dc992/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L209) the custom `struct` which represents the `BlockNode` and its `DFSNum`

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,659 @@
+template<class GraphT, class GT>
+void CycleEquivalenceAnalysis<GraphT, GT>::insertCappingBackedges(BlockGraph
+                                                                    &Graph) {
+  // Highest, i.e., closer to the spanning tree root, reached DFSNum for each
+  // `BB`
```
Please add that for nodes that are "higher", so close to the root, the `DFSNum` is smaller. Otherwise the rest gets confusing. At least if you say it out loud in a comment there's a place to refer to when you're confused, and it doesn't seem wrong to use `std::min` all over the place to compute the "highest".

BTW why did you decide to use the "highest" on your own or to use the same terminology of the paper?

If you did it for the paper, keep it like this but explain that's the reason. It can still get confusing but if you read the paper everything makes sense. 
Instead, if it was your choice please revert it all over the codebase because it can get very confusing for no actual reason.

andrealinux1 replied:
The terminology is taken from the paper, I tried to keep it as coherent as possible with it. I added here https://github.com/revng/revng-c/blob/ce352f2912e639e1ef307bb1d5b131db48444878/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L178 the disclaimer on top on the explanation "highest -> lower dfsnum", let me know if you still want to change the use of the terms.

pfez replied:
When you talk about "the paper" state what paper you're talking about.

You can put a big comment at the beginning of `CycleEquivalenceAnalysis.cpp` explaining that it contains the implementation of the paper, with a full citation of the paper, and possibly a link to the PDF. In this way in all the other places in the code where you talk about "the paper" it's clear which one.

andrealinux1 replied:
I already had a comment before the `CycleEquivalenceAnalasysis` class [here](https://github.com/revng/revng-c/blob/f23593da9f15288aca2fee41975e6e55dbbbcb3b/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L119), but I also added a small reference, as you suggested, [here](https://github.com/revng/revng-c/blob/f23593da9f15288aca2fee41975e6e55dbbbcb3b/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L5) at the beginning of the `cpp` file.

pfez replied:
I remember there wasn't one, but I couldn't find it during review. It must have been an off-by-grep. Thanks for adding it anyways.

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,659 @@
+      DFSNumNodeMap[Target->getDFSNum()] = Target;
+    }
+
+    // 2) Compute the highest reached node for any child of the current `Block`
+    size_t HI1 = SizeTMaxValue;
+    for (const auto &[ForwardEdge, _, __] : getForwardEdges(BB)) {
+      BlockNode *Target = std::get<1>(ForwardEdge);
+      HI1 = std::min(HI1, HighestReachedDFSNum.at(Target));
+    }
+
+    // 3) Assign the highest reached node for the current `Block`
+    HighestReachedDFSNum[BB] = std::min(HI0, HI1);
+
+    // 4) Compute the child which has the reached DFSNum computed at step 2)
+    BlockNode *HI1Child = nullptr;
+    if (HI1 != SizeTMaxValue) {
+      for (const auto &[ForwardEdge, _, __] : getForwardEdges(BB)) {
+        BlockNode *Target = std::get<1>(ForwardEdge);
+        if (HI1 == HighestReachedDFSNum.at(Target)) {
+          HI1Child = Target;
+          break;
+        }
+      }
+    }
```
I actually think that you can merge 2 and 4 in a single loop, very similar to what I proposed for point 5.

And do point 3 after old point 2+4(merged) and before point 5.

andrealinux1 replied:
Conversely to the previous, point, the 2 and 4 steps merge is actually doable, and I did it here https://github.com/revng/revng-c/blob/b5fe805f75dde9c9ada264ace247d4b989c4f2d7/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L213.

I did not do it previously because I wanted to keep the structure of the code as similar as possible to how it is laid out in the paper.

pfez commented on lib/RestructureCFG/CycleEquivalenceAnalysis.cpp:
```diff
@@ -0,0 +1,659 @@
+template<class GraphT, class GT>
+void CycleEquivalenceAnalysis<GraphT, GT>::insertEdge(BlockEdgeDescriptor E,
+                                                      EdgeLabel *Label,
+                                                      bool IsInverted,
```
It seems a little bit awkward to compute `IsInverted` very early (in `getTreeEdges`, `getBackwardEdges` ...) and make it propagate until here.

I think that `getTreeEdges` and friends could avoid to compute that and just return edge iterators that point to the `Undirected` edges.

Then, nobody else uses the `IsInverted` field, and the edge iterators end up being passed in here instead of `BlockEdgeDescriptor` and `EdgeLabel` (both can be recovered from the edge iterator IIUC).

So the computation of `IsInverted` can be centralized here, which is the only place where we care about I think. And perhaps you can even drop the `BlockEdgeDescriptor` alltogether.

andrealinux1 replied:
Mmmh, I have some doubt on how to implement this though:

You suggested returning the edge iterators from `getTreeEdges`, and I also think that it would be okay in all the call sites, except in the `insertEdge` method, where I actually need the `IsInverted` property.

Indeed, how can I compute the `IsInverted` here, just from the edge iterators?
Assume that I'm in a invocation of the `insertEdge`, I get as a parameter the edge iterator, and I want to compute if the passed edge iterator refers to a `Inverted` edge. Doing such thing would mean checking if the current edge, when we "see" the graph as `Undirected` is part of the "predecessor" half, meaning it has surpassed the "successor" half (remember that we `concat(successors + predecessors)`. I cannot compare the edge iterator passed by parameter with the iterator of the last edge which is not inverted not applying the `Undirected` trait.

But even admitting that I pass both the edge iterator of the edge we want to insert, and the iterator pointing to the last edge which is not inverted (and I would need to return both such iterators from the `getTreeEdges`, which is not super elegant), I cannot compare with `<` iterators that are not `random_access_iterator`s.

Therefore, the only way to compute the `IsInverted`, is to start from zero again and iterate over all the successors every time we invoke the `insertEdge` method, so we can keep track of the `IsInverted` property using the `Index` of the successors as it is done now in the `getTreeEdges` method, which doesn't seem nice and not what you are suggesting.

What am I missing and not understanding with respect to what you suggested?

pfez replied:
Yeah your line of reasoning makes sense. Thanks for walking me through it.

pfez commented:
NOTE TO SELF: everything above here is on an older version of the patchset. Re-review it before merging to make sure I haven't lost anything.

pfez commented on include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h:
```diff
@@ -0,0 +1,257 @@
+
+/// The `EdgeLabel` `struct` is used to represent the information needed to
+/// identify each undirected edge in the `GenericGraph` materialized internally
+/// in the analysis, and to map it eventually on the edge on the undirected
```
I think you meant directed here, not undirected.

andrealinux1 replied:
Right, it was a typo, corrected it [here](https://github.com/revng/revng-c/blob/f23593da9f15288aca2fee41975e6e55dbbbcb3b/include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h#L34)
